File: python_open/hough.py
import cv2
import numpy as np

# 이미지 경로 설정
folder = "/home/test/Desktop/opencv/opencv/cppOpen/build/data/"
src = cv2.imread(folder + "img.bmp", cv2.IMREAD_GRAYSCALE)

if src is None:
    print("Error: Could not load image!")
    exit(-1)

# Hough 변환을 위한 이진 이미지 생성
_, binary = cv2.threshold(src, 128, 255, cv2.THRESH_BINARY)

# Hough 변환으로 선 검출
lines = cv2.HoughLines(binary, 1, np.pi / 180, 100)

# 선을 그릴 이미지 생성
line_image = np.zeros((src.shape[0], src.shape[1], 3), dtype=np.uint8)

# Hough 선 그리기
if lines is not None:
    for line in lines:
        rho, theta = line[0]  # rho와 theta 추출
        a = np.cos(theta)
        b = np.sin(theta)
        x0 = a * rho
        y0 = b * rho
        pt1 = (int(x0 + 1000 * (-b)), int(y0 + 1000 * (a)))
        pt2 = (int(x0 - 1000 * (-b)), int(y0 - 1000 * (a)))
        cv2.line(line_image, pt1, pt2, (0, 0, 255), 1, cv2.LINE_AA)  # 선 그리기

# 결과 표시
cv2.imshow("Original Image", src)
cv2.imshow("Hough Lines", line_image)

cv2.waitKey(0)  # 키 입력 대기
cv2.destroyAllWindows()

File: cppOpen/hough.cpp
#include "opencv2/opencv.hpp"
#include <iostream>

using namespace cv;
using namespace std;

String folder = "/home/test/Desktop/opencv/opencv/cppOpen/build/data/";

int main() {
    // 이미지 읽기
    Mat src = imread(folder + "img.bmp", IMREAD_GRAYSCALE);
    if (src.empty()) {
        cout << "이미지를 열 수 없거나 찾을 수 없습니다!" << endl;
        return -1;
    }

    // Hough 변환을 위한 이진 이미지 생성
    Mat binary;
    threshold(src, binary, 128, 255, THRESH_BINARY);

    // Hough 변환으로 선 검출
    vector<Vec2f> lines;
    HoughLines(binary, lines, 1, CV_PI / 180, 100);

    // 선을 그릴 이미지 생성
    Mat lineImage = Mat::zeros(src.size(), CV_8UC3);

    // Hough 선 그리기
    for (const auto& lineP : lines) {
        float rho = lineP[0];
        float theta = lineP[1];
        Point pt1, pt2;
        double a = cos(theta), b = sin(theta);
        double x0 = a * rho, y0 = b * rho;
        pt1.x = cvRound(x0 + 1000 * (-b));
        pt1.y = cvRound(y0 + 1000 * (a));
        pt2.x = cvRound(x0 - 1000 * (-b));
        pt2.y = cvRound(y0 - 1000 * (a));
        line(lineImage, pt1, pt2, Scalar(0, 0, 255), 1, LINE_AA); // 선 그리기
    }

    // 결과 표시
    imshow("Original Image", src);
    imshow("Hough Lines", lineImage);

    waitKey(0); // 키 입력 대기
    return 0;
}
